node/src/utils/median.rs

```diff
@@ -0,0 +1,55 @@
+use std::{
+    collections::BTreeMap,
+    ops::{Add, AddAssign},
+};
+
+pub(crate) fn weighted_median<T, I, Weight>(iterator: I) -> Option<T>
+where
+    T: Ord,
+    Weight: Default + Add<Weight, Output = Weight> + AddAssign<Weight> + PartialOrd + Copy,
+    I: IntoIterator<Item = (T, Weight)>,
+{
+    let (value_weights, total_weight): (_, Weight) = iterator.into_iter().fold(
+        (BTreeMap::<T, Weight>::new(), Default::default()),
+        |(mut weights, total), (value, weight)| {
+            *weights.entry(value).or_default() += weight;
+            (weights, total + weight)
+        },
+    );
+    let mut counted: Weight = Default::default();
+    let mut result = None;
+    for (value, weight) in value_weights {
+        result = Some(value);
+        counted += weight;
+        if counted + counted >= total_weight {
+            break;
```

Comment:
Why not just `return Some(value);` here? Then you don't need the `result` variable.